webhook/webhook.go

```diff
@@ -0,0 +1,102 @@
+package webhook
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	log "github.com/gophish/gophish/logger"
+)
+
+const (
+
+	// DefaultTimeoutSeconds is amount of seconds of timeout used by HTTP sender
+	DefaultTimeoutSeconds = 10
+
+	// MinHTTPStatusErrorCode is the lowest number of an HTTP response which indicates an error
+	MinHTTPStatusErrorCode = 400
+
+	// SignatureHeader is the name of an HTTP header used to which contains signature of a webhook
+	SignatureHeader = "X-Gophish-Signature"
+)
+
+// Sender defines behaviour of an entity by which webhook is sent
+type Sender interface {
+	Send(endPoint EndPoint, data interface{}) error
+}
+
+type defaultSender struct {
+	client *http.Client
+}
+
+var senderInstance = &defaultSender{
+	client: &http.Client{
+		Timeout: time.Second * DefaultTimeoutSeconds,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	},
+}
+
+// EndPoint represents and end point to send a webhook to: url and secret by which payload is signed
+type EndPoint struct {
+	URL    string
+	Secret string
+}
+
+// Send sends data to a single EndPoint
+func Send(endPoint EndPoint, data interface{}) error {
+	return senderInstance.Send(endPoint, data)
+}
+
+// SendAll sends data to each of the EndPoints
+func SendAll(endPoints []EndPoint, data interface{}) {
+	for _, ept := range endPoints {
+		go func(ept1 EndPoint) {
+			senderInstance.Send(ept1, data)
+		}(EndPoint{URL: ept.URL, Secret: ept.Secret})
+	}
+}
+
+// Send contains the implementation of sending webhook to an EndPoint
+func (ds defaultSender) Send(endPoint EndPoint, data interface{}) error {
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	req, err := http.NewRequest("POST", endPoint.URL, bytes.NewBuffer(jsonData))
+	signat, err := sign(endPoint.Secret, jsonData)
```

Comment:
After reviewing some of the common webhook server implementations, I think we should tweak this implementation slightly. Namely, we should add a prefix to the signature to indicate what hashing algorithm was used. This would make it like:

```
X-Gophish-Signature: sha256=hash
```

You can see an example of this [here](https://developer.github.com/webhooks/#example-delivery). The benefit this gives us is making it much easier to move to different hashing functions in the future as needed.

Reply:
For now, we can keep this simple - I'd implement this is to make a new constant, `Sha256Prefix = "sha256"`, and then adjust the signature that's set.

Reply:
fixe